### PR TITLE
Add synthetic dataset, test for simple training with synthetic data

### PR DESCRIPTION
**Pull request by achalddave**
This allows us to do simple tests that call through main(), similar to open_clip: https://github.com/mlfoundations/open_clip/blob/main/tests/test_training_simple.py

### CODE REVIEW

**Comment by mitchellnw** [comment]
LGTM

**Review by mitchellnw — APPROVED** [review]

File: tests/test_training_simple.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    batch_size = 1
+    # fmt: off
+    main([
+        "--train-num-samples", str(num_batches * seq_len),
+        "--batch-size", str(batch_size),
+        "--dataset-type", "synthetic",
+        "--model", "open_lm_11m",
+        "--epochs", "1",
+    ])
+    # fmt: on
```
Comment (sagadre) [inline tests/test_training_simple.py:16]: probably need something like:
`success = main(main_args)`
`assert success != -1`

Comment (achalddave) [reply tests/test_training_simple.py:16]: main() should throw an exception if anything breaks, which will cause the test to fail. This is probably ideal for  tests, running from command line, and for development (so you don't need to think about returning 0/1 from main and just throw an exception instead). What do you think?

Comment (sagadre) [reply tests/test_training_simple.py:16]: agreed, though right now main returns -1 in many failure cases. so we can either add throwing to this PR or check for -1?

Comment (achalddave) [reply tests/test_training_simple.py:16]: oh, didn't even realize we did that. I'll add throwing to this PR.

Comment (achalddave) [reply tests/test_training_simple.py:16]: added! changed all `return -1` to `raise ValueError` instead

Comment (sagadre) [reply tests/test_training_simple.py:16]: nice! lgtm!

File: tests/test_training_simple.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+def test_train_simple():
+    seq_len = 2048
+    num_batches = 5
+    batch_size = 1
+    # fmt: off
+    main([
+        "--train-num-samples", str(num_batches * seq_len),
+        "--batch-size", str(batch_size),
+        "--dataset-type", "synthetic",
+        "--model", "open_lm_11m",
```
Comment (iejMac) [inline tests/test_training_simple.py:13]: maybe we should also add a "test config" which is even smaller? Like width is fine but maybe let's make it a 2-3 layer transformer? Call it `open_lm_test_config.json` or sth

Comment (achalddave) [reply tests/test_training_simple.py]: good idea, added

File: open_lm/model_configs/open_lm_test_tiny.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+    "hidden_dim": 96,
+    "n_layers": 2,
+    "n_heads": 2,
+    "seq_len": 2048,
+    "vocab_size": 50432,
```
Comment (iejMac) [inline open_lm/model_configs/open_lm_test_tiny.json:6]: being a bit tedious here so feel free to ignore but couldn't we go down to some other setting for testing - like character level tokenization (would make vocab_size much smaller) and shorter ctx len?

Right now even with hidden dim 96 you will have at least `96 * 50432 = 4841472` embedding params and this last matmul for the output probs will be slow

Comment (iejMac) [reply open_lm/model_configs/open_lm_test_tiny.json:6]: this is obv fine for now if it works

Comment (achalddave) [reply open_lm/model_configs/open_lm_test_tiny.json]: not tedious at all, makes sense, especially for this test. made everything smaller

**Comment by iejMac** [comment]
LGTM! Unit tests pass in <3min which is super nice

**Pull request merged**